documentation/public/internals/design-docs/language-definition-v2/index.md

```diff
@@ -0,0 +1,438 @@
+Their fields match 1-1 with the item fields. The struct name contributes a `NonTerminalKind`.
+
+Each field can be either `Required(T)` or `Optional(T)`. Required fields are always present and parsed. Optional fields
+can be omitted if they don't exist, and are represented with Rust's `Option<T>` type (or TypeScript `T | undefined`).
+However, optional fields have additional `enabled_in` and `disabled_in` properties. After parsing optional fields,
+we should compare them with the current language version, and produce errors if they don't match, but continue parsing normally.
+
+The type of each field can be a `NonTerminal(T)` or `Terminal(Set<T>)`. A non-terminal field refers to another item, and
+holds its type. A terminal field refers to one or more terminal items (all valid in this position), and is of type `TerminalNode`.
+
+Additionally, the struct also stores the CST node that holds its contents. The idea is for the parser to attempt
+constructing both the CST and the AST nodes in one pass.
+
+```rust title="Definition"
+Struct(
+    name = ParametersDeclaration,
+    fields = (
+        open_paren = Required(Terminal([OpenParen])),
+        parameters = Required(NonTerminal(Parameters)),
+        close_paren = Required(Terminal([CloseParen]))
+    )
+)
+```
+
+```rust title="AST Type"
+pub struct ParametersDeclaration {
+    pub open_paren: Rc<TerminalNode>,
+    pub parameters: Rc<Parameters>,
+    pub close_paren: Rc<TerminalNode>,
+
+    pub cst: Rc<NonTerminalNode>,
+}
+```
+
+```rust title="Parser Pseudo-Code"
+fn parse_parameters_declaration(input) -> ParametersDeclaration {
+    let mut cst = NonTerminalNode::new(NonTerminalKind::ParametersDeclaration);
+
+    // Parses any leading trivia, the 'OpenParen' token, and any trailing trivia
+    // Add all of them to the 'cst' parent node (in order)
+    // Then return a reference to the 'OpenParen' token
+    let open_paren = parse_terminal(input, TerminalKind::OpenParen, &mut cst);
+
+    // Parses and returns the 'Parameters' AST node.
+    let parameters = parse_parameters(input);
+    cst.children.push(parameters.cst.clone());
+
+    // Parses any leading trivia, the 'CloseParen' token, and any trailing trivia
+    // Add all of them to the 'cst' parent node (in order)
+    // Then return a reference to the 'CloseParen' token
+    let close_paren = parse_terminal(input, TerminalKind::CloseParen, &mut cst);
+
+    return ParametersDeclaration {
+        open_paren,
+        parameters,
+        close_paren,
+
+        cst,
+    };
+}
+```
+
+### Enum Items
+
+Enums represent an ordered choice of multiple variants (possibilities). The enum name contributes a `NonTerminalKind`,
+but not its variants (they only exist in the AST, and don't affect the CST at all).
+
+We attempt to parse each variant (in-order), and choose the first one that succeeds. However, each variant can have
```

Comment:
i.e. formally this is an `Ordered Choice Operator`

Reply:
added.